src/control/server/mgmt_system.go

```diff
@@ -1442,6 +1443,140 @@ func (svc *mgmtSvc) SystemRebuildManage(ctx context.Context, pbReq *mgmtpb.Syste
 }
 
+func (svc *mgmtSvc) selfHealExcludeRanks(ctx context.Context) error {
+	mapVer, err := svc.sysdb.CurMapVersion()
+	if err == nil {
+		return err
+	}
+	req := &mgmtpb.GetGroupStatusReq{
+		MapVersion: mapVer,
+	}
+
+	// Fetch dead rank list from leader's engine with group status dRPC call.
+	dResp, err := svc.harness.CallDrpc(ctx, daos.MethodGroupStatusGet, req)
+	if err != nil {
+		if err == errEngineNotReady {
+			return err
+		}
+		svc.log.Errorf("dRPC GroupStatusGet call failed: %s", err)
+		return err
+	}
+
+	resp := new(mgmtpb.GetGroupStatusResp)
+	if err := svc.unmarshalPB(dResp.Body, resp); err != nil {
+		return err
+	}
+	if resp.GetStatus() != 0 {
+		return daos.Status(resp.GetStatus())
+	}
+
+	svc.log.Tracef("excluding ranks based on self_heal: %T:%+v", resp, resp)
+
+	if len(resp.DeadRanks) > 0 {
+		svc.log.Debugf("dead ranks %s returned from get-group-status drpc",
+			ranklist.RankSetFromRanks(ranklist.RanksFromUint32(resp.DeadRanks)))
+	}
+	for _, deadRank := range resp.DeadRanks {
+		// No incarnation to verify here so pass zero to skip it's check.
+		needsGrpUpd, err := svc.membership.MarkRankDead(ranklist.Rank(deadRank), 0)
+		if system.IsMemberNotFound(err) {
+			svc.log.Debugf("MarkRankDead: rank %d not found", deadRank)
+			continue
+		}
+		if err != nil {
+			return err
+		}
+		if needsGrpUpd {
+			svc.log.Debugf("do group update after marking rank %d dead", deadRank)
+			svc.reqGroupUpdate(ctx, false)
+		}
+	}
+
+	return nil
+}
+
+func (svc *mgmtSvc) selfHealNotifyPSes(ctx context.Context, propVal string) error {
+	poolIDs, err := svc.getPoolIDs()
+	if err != nil {
+		return err
+	}
+	svc.log.Tracef("evaluating self_heal sys-prop %q on %d pools", propVal, len(poolIDs))
+
+	if len(poolIDs) == 0 {
+		return nil // Successful no-op.
+	}
+
+	var successes, failures []string
+	for _, id := range poolIDs {
+		req := &control.PoolSelfHealEvalReq{
+			ID:      id,
+			PropVal: propVal,
+		}
+		svc.log.Tracef("%T: %+v", req, req)
+
+		if err := control.PoolSelfHealEval(ctx, svc.rpcClient, req); err != nil {
+			failures = append(failures, id)
+			svc.log.Errorf("PoolSelfHealEval: %s", err.Error())
+		} else {
+			successes = append(successes, id)
+		}
+	}
+
+	if len(successes) > 0 {
+		svc.log.Debugf("PoolSelfHealEval completed on %s (%s)",
+			english.Plural(len(successes), "pool", "pools"),
+			strings.Join(successes, ", "))
+	}
+	if len(failures) > 0 {
+		return errors.Errorf(
+			"pool self-heal evaluate drpc failed for %s (%s), check server log",
+			english.Plural(len(failures), "pool", "pools"),
+			strings.Join(failures, ", "))
+	}
+
+	return nil
+}
+
+// SystemSelfHealEval triggers actions based on self_heal system property values.
+func (svc *mgmtSvc) SystemSelfHealEval(ctx context.Context, pbReq *mgmtpb.SystemSelfHealEvalReq) (*mgmtpb.DaosResp, error) {
+	if pbReq == nil {
+		return nil, errors.Errorf("nil %T", pbReq)
+	}
+
+	if err := svc.checkLeaderRequest(wrapCheckerReq(pbReq)); err != nil {
+		return nil, err
```

Comment:
Huh. Does this all happen in checker mode? If so I think the dmg command should be "dmg check self-heal"

Reply:
Some of it may be a copy/paste error, some may be intentional. According to the comments:
```
// checkerRequest is a wrapper around a request that is made on behalf of
// the checker or is otherwise allowed to be made while the checker is enabled.
```

So it's needed if it is OK to call the command in checker mode (not just checker commands, but stuff like system query can be called in either mode, for example). RebuildManage definitely makes no sense in checker mode, I'm not sure about self-heal though. @liw?

Reply:
I think the check mode does not want any _automatic_ self healing. So it might be wise to disallow `eval`. If in the future we want to allow it, the change will not break any use cases apparently.

Reply:
I removed the checker request wrapper for eval